### PR TITLE
Include composer.lock in git files

### PR DESCRIPTION
**Pull request by ArthurHoaro**
When `composer update` is run, Composer will retrieve latest dependencies and generate a new `composer.lock`. With `composer install`, it will retrieve the exact same version present in `composer.lock`, i.e. the latest `composer update` versions.

This should prevent us from running in the same issue we did with #779 

TODO: 
  - [x] update the docs.

### CODE REVIEW

**Comment by nodiscc** [comment]
Sorry I didn't follow #779 too closely. What was the problem and how does this patch fix it? (I guess it related to not having to manually update dependency version numbers?)

**Comment by ArthurHoaro** [comment]
I released a version of netscape bookmark parser in `v1.x` with a breaking backward compatibility change. Without `composer.lock`, any composer grab had resulted with an incompatible dependency (because it was fetching `1.*` version).

This PR isn't really fixing it because I still would have to remove the breaking change, but with `composer install` and `composer.lock` it wouldn't have been critical because CI and new installs would still be working.

I'm not sure that my explanation is very clear, feel free to ask if not.

**Comment by virtualtam** [comment]
@nodiscc lock files allow pinning dependencies for end-users so applications are set up in a controlled environment, while using a loose configuration for development (`composer.json`). The drawback is that we'll have to manually update it.

Related resources:
- https://stackoverflow.com/questions/27016865/gemfile-lock-equivalent-in-maven-and-gradle
- https://bundler.io/v1.3/rationale.html#checking-your-code-into-version-control
- https://stackoverflow.com/questions/10674641/composer-lock-how-does-it-work

File: composer.lock [hunk 1]
```diff
@@ -0,0 +1,2372 @@
+{
```
Comment (virtualtam) [inline composer.lock]: shouldn't the lock file be generated in a production (vs. development) environment?

File: .gitattributes [hunk 2]
```diff
@@ -29,3 +29,4 @@ Doxyfile        export-ignore
 Makefile        export-ignore
 phpunit.xml     export-ignore
 tests/          export-ignore
+composer.lock   export-ignore
```
Comment (virtualtam) [inline .gitattributes]: - the lock file should be included in release archives so we have reproducible Docker builds
- it should be considered by Git as a JSON file for Github to show diffs during reviews (else we'll probably miss involuntary changes)

File: composer.lock (hunk 1 above)
Comment (ArthurHoaro) [inline composer.lock]: It doesn't matter, it supports `--no-dev` parameter.

File: .gitattributes (hunk 2 above)
Comment (ArthurHoaro) [inline .gitattributes]: Ok, you're right.

**Comment by ArthurHoaro** [comment]
> The drawback is that we'll have to manually update it.

Yup, that's not really an issue. We can do this from time to time or when we need a specific release.

File: composer.lock (hunk 1 above)
Comment (virtualtam) [inline composer.lock]: 👍

**Comment by ArthurHoaro** [comment]
Rebased.

**Pull request merged**